src/lib.rs

```diff
@@ -140,127 +117,142 @@ impl<I: core::iter::Iterator> Peekable for core::iter::Peekable<I> {
 pub trait Miniconf {
-    /// Update an element by path.
+    /// Deserialize an element by path.
     ///
     /// # Args
-    /// * `path` - The path to the element with '/' as the separator.
-    /// * `data` - The serialized data making up the content.
+    /// * `path_parts`: A [Peekable] [Iterator] identifying the element.
+    /// * `de`: A [serde::Deserializer] to use to deserialize the value.
     ///
     /// # Returns
-    /// The number of bytes consumed from `data` or an [Error].
-    fn set(&mut self, path: &str, data: &[u8]) -> Result<usize, Error> {
-        self.set_path(&mut path.split('/').peekable(), data)
-    }
+    /// May return an [Error].
+    fn set_path<'a, 'b: 'a, P, D>(&mut self, path_parts: &mut P, de: D) -> Result<(), Error>
+    where
+        P: Peekable<Item = &'a str>,
+        D: serde::Deserializer<'b>;
 
-    /// Retrieve a serialized value by path.
+    /// Serialize an element by path.
     ///
     /// # Args
-    /// * `path` - The path to the element with '/' as the separator.
-    /// * `data` - The buffer to serialize the data into.
+    /// * `path_parts`: A [Peekable] [Iterator] identifying the element.
+    /// * `ser`: A [serde::Serializer] to use to serialize the value.
     ///
     /// # Returns
-    /// The number of bytes used in the `data` buffer or an [Error].
-    fn get(&self, path: &str, data: &mut [u8]) -> Result<usize, Error> {
-        self.get_path(&mut path.split('/').peekable(), data)
-    }
+    /// May return an [Error].
+    fn get_path<'a, P, S>(&self, path_parts: &mut P, ser: S) -> Result<S::Ok, Error>
+    where
+        P: Peekable<Item = &'a str>,
+        S: serde::Serializer;
+
+    /// Get the next path in the namespace.
+    ///
+    /// This is usually not called directly but through a [MiniconfIter] returned by [SerDe::iter_paths].
+    ///
+    /// # Args
+    /// * `state`: A state array indicating the path to be retrieved.
+    ///   A zeroed vector indicates the first path. The vector is advanced
+    ///   such that the next element will be retrieved when called again.
+    ///   The array needs to be at least as long as the maximum path depth.
+    /// * `path`: A string to write the path into.
+    /// * `separator` - The path hierarchy separator.
+    ///
+    /// # Returns
+    /// A `bool` indicating a valid path was written to `path` from the given `state`.
+    /// If `false`, `path` is invalid and there are no more paths within `self` at and
+    /// beyond `state`.
+    /// May return `IterError` indicating insufficient `state` or `path` size.
+    fn next_path<const TS: usize>(
+        state: &mut [usize],
+        path: &mut heapless::String<TS>,
+        separator: char,
+    ) -> Result<bool, IterError>;
+
+    /// Get metadata about the paths in the namespace.
+    fn metadata() -> Metadata;
+}
+
+pub trait SerDe<S>: Miniconf {
+    /// The path hierarchy separator.
+    const SEPARATOR: char;
```

Comment:
Mind adding a note that this separator type is used by the path iterators? I couldn't figure out why it was a member of the trait for a long time, and we should document the usage of it for docs.rs